docs(tutorial): add new tutorial covering core features (#769)
awslabs/aws-lambda-powertools-python: docs/quickstart.md

---
title: QuickStart
description: Powertools introduction
---
Quickstart introducing core Powertools functionalities.

## Installation
With [SAM](https://docs.aws.amazon.com/serverless-application-model/latest/developerguide/serverless-sam-cli-install.html) installed, you can either follow this quickstart step by step or you can create project with final quickstart code to play with.

> If you want to follow it from very beginning just create new empty project.

Author: done

=== "shell"
```bash
sam init --runtime python3.9 --dependency-manager pip --app-template hello-world --name powertools-quickstart
```
If you want to play with the final code then you can download it from [aws-samples](https://github.com/aws-samples/cookiecutter-aws-sam-python) github repository.
=== "shell"
```bash
sam init --location https://github.com/aws-samples/cookiecutter-aws-sam-python
```

### Code example and Configuration
Inside of the powertools-quickstart directory we will modify following files:

> * **app.py** - File that contains our application code.

Author: done

* **template.yaml** - AWS infrastructure configuration using SAM template format.
* **requirements.txt** - Python pip packages configuration.

Also, ensure you [configure your AWS credentials](https://docs.aws.amazon.com/serverless-application-model/latest/developerguide/serverless-getting-started-set-up-credentials.html) to be able to connect and deploy example code into your AWS account. 

> Once it is done, lets configure our basic application to looks like the following code snippet.

Author: moved higher as suggested

=== "app.py"

    ```python
    def hello():
        return {"statusCode": 200, "body": {"message": "hello unknown!"}}

    def lambda_handler(event, context):
        return hello()
    ```

=== "template.yaml"

    ```yaml
    AWSTemplateFormatVersion: '2010-09-09'
    Transform: AWS::Serverless-2016-10-31
    Description: >
        Sample SAM Template for powertools-quickstart
    Globals:
    Function:
        Timeout: 3
    Resources:
    HelloWorldFunction:
        Type: AWS::Serverless::Function
        Properties:
        CodeUri: hello_world/
        Handler: app.lambda_handler
        Runtime: python3.9
        Architectures:
            - x86_64
        Events:
            HelloWorld:
            Type: Api 
            Properties:
                Path: /hello
                Method: get
    Outputs:
        HelloWorldApi:
            Description: "API Gateway endpoint URL for Prod stage for Hello World function"
            Value: !Sub "https://${ServerlessRestApi}.execute-api.${AWS::Region}.amazonaws.com/Prod/hello/"
    ```

> Our lambda code consist of `lambda_handler` method which is invoked by API and `hello` method which returns result to the API Gateway and is invoked by handler itself.

Author: correct, fixed

SAM template configures API Gateway that redirects traffic to Lambda for only one path: `hello`. 
!!! Warning 
    For simplicity we don't configure authentication and authorization in the example!
### Run your code

> On every step you have two ways of testing your code. Locally and in your AWS account. Since we use SAM both methods are equally straightforward.

Author: You are right, lets use 'running your code' in this section as it closer to reality than testing

Value               https://1234567890.execute-api.eu-central-1.amazonaws.com/Prod/hello/                                                

Key                 HelloWorldFunction                                                                                                   
Description         Hello World Lambda Function ARN                                                                                      
Value               arn:aws:lambda:eu-central-1:123456789012:function:sam-app-HelloWorldFunction-dOcfAtYoEiGo                           
------------------------------------------------------------------------------------------------------------------------------------------
Successfully created/updated stack - sam-app in eu-central-1
```
This command will build package and deploy it to your AWS Account. Make sure you configure AWS credentials beforehand. In the outputs section you find API Url you can issue request against.
!!! Info
    To learn more about SAM deployment mechanism follow this [link](https://docs.aws.amazon.com/serverless-application-model/latest/developerguide/sam-cli-command-reference-sam-deploy.html).

## API Gateway router

> Let's say you have been asked to create another method that respond with caller name under `/hello/<name>` URL path.

Reviewer: Is that an echo method? If yes, maybe we can first describe the problem (return input) and then the implementation (use path variable in url. Suggestion:

Let's say we want to have a method that outputs a string that we send back, like an echo. We would need to create an API with an URL path `/hello/<message>`, where the message is the input from the user.

Author: Added something like this:
```Let's say we want to have another method that behaves as an echo. It takes user input (username) and output it to the caller. We would need to create an API with an URL path /hello/<name>, where the name string is the input from the user.
```

    Transform: AWS::Serverless-2016-10-31
    Description: > 
        Sample SAM Template for powertools-quickstart
    Globals:
        Function:
            Timeout: 3
    Resources:
        HelloWorldFunction:
            Type: AWS::Serverless::Function
            Properties:
            CodeUri: hello_world/
            Handler: app.lambda_handler
            Runtime: python3.9
            Events:
                HelloWorld:
                Type: Api 
                Properties:
                    Path: /hello
                    Method: get
        HelloWorldFunctionName:
            Type: AWS::Serverless::Function
            Properties:
            CodeUri: hello_world/
            Handler: app_name.lambda_handler
            Runtime: python3.9
            Events:
                HelloWorldName:
                Type: Api
                Properties:
                    Path: /hello/{name}
                    Method: get
    Outputs:
        HelloWorldApi:
            Description: "API Gateway endpoint URL for Prod stage for Hello World function"
            Value: !Sub "https://${ServerlessRestApi}.execute-api.${AWS::Region}.amazonaws.com/Prod/hello/"
    ```

> This way definitively works. The downside of this approach is that you need to create new lambda file for every method or path and also configure SAM template to add new lambda configuration and configure API Gateway to point into them. For small additional methods it might be too much boilerplate code.  

Reviewer: This sound a bit biased, where we present a downside without having a problem or complexity. Maybe first describe the situation when we need to add a new method, and that it will be too much boilerplate for many URL paths?

Author: rewrote to:
`This way certainly works for simple use-case. But what happens if your application gets bigger and we need to cover numerous URL paths and HTTP methods for them? If that is the case, we should:
* Add a new lambda handler with business logic for each new URL path and HTTP method used.
* Add a new Lambda configuration to a SAM template file to map the lambda function to the required path and HTTP URL method.
This could result in a number of similar lambda files and large SAM configuration file with similar configuration sections. 
In this case where we see that the addition of new URL paths lead to the boilerplate code, we should lean towards the routing approach.`

> Another approach though that might be more suitable in this case is to have all of our methods under control of one lambda and "route" into specific method based on a request. 

Reviewer: This can be much shorter and straight to the point: If we have many methods we should use routing.

Author: added based on suggestion

Example, naive code might look like the following code snippet.
=== "app.py"

    ```python hl_lines="10-23 26-28 34-35"
        def hello_name(event, **kargs):
            username = event["pathParameters"]["name"]
            return {"statusCode": 200, "body": {"message": f"hello {username}!"}}

        def hello(**kargs):
            return {"statusCode": 200, "body": {"message": "hello unknown!"}}

        class Router:
            def __init__(self):
                self.routes = {}

            def set(self, path, method, handler):
                self.routes[f"{path}-{method}"] = handler

            def get(self, path, method):
                try:
                    route = self.routes[f"{path}-{method}"]
                except:
                    print("Cannot route request to correct method")
                    raise NotImplemented
                return route

        router = Router()
        router.set(path="/hello", method="GET", handler=hello)
        router.set(path="/hello/{name}", method="GET", handler=hello_name)

        def lambda_handler(event, context):
            path = event["resource"]

>             http_method = event["httpMethod"]

Reviewer: no error handling here, not sure if intentional, if yes, just ignore my comment.

Author: this is on purpose to keep it simple

    ```

=== "template.yaml"

    ```yaml hl_lines="18-28"
        AWSTemplateFormatVersion: "2010-09-09"
        Transform: AWS::Serverless-2016-10-31
        Description: >
            Sample SAM Template for powertools-quickstart
        Globals:
            Function:
            Timeout: 3
        Resources:
            HelloWorldFunction:
            Type: AWS::Serverless::Function
            Properties:
                CodeUri: hello_world/
                Handler: app.lambda_handler
                Runtime: python3.9
                Events:
                    HelloWorld:
                        Type: Api
                        Properties:
                        Path: /hello
                        Method: get
                    HelloWorldName:
                        Type: Api 
                        Properties:
                        Path: /hello/{name}
                        Method: get
        Outputs:
            HelloWorldApi:
                Description: "API Gateway endpoint URL for Prod stage for Hello World function"
                Value: !Sub "https://${ServerlessRestApi}.execute-api.${AWS::Region}.amazonaws.com/Prod/hello/"
    ```

> In this example we added two methods `hello_name` and `hello` next to each other. We added `Router` class that allow us to register what method should be called when specific request comes. 

Reviewer: simplify: ~~next to each other~~

Author: Rephrased to:
`In this example, two `hello_name` and `hello` methods were added. We added the `Router` class which allows us to record the method that should be called when the specific request arrives. `

We created instance of it and added configuration with mapping between processing methods and http request method. 
In lambda handler we call router instance get method to fetch reference to processing method (`hello` or `hello_name`) that will handle the request. 

> Finally we run this method and return result to API Gateway.

Reviewer: It's a bit hard to read, maybe we can reference the line numbers of the code or explicitly mention specific code parts again?

Author: Added enumeration and line information. Let me know if this is enough

Thanks to this approach, we can simplify our infrastructure configuration since we added all API Gateway path under `HelloWorldFunction` event section. 

> Downside of current approach is that we need to add custom class and with every new method or path we want to serve we need to add explicit configuration. 

Reviewer: The downside is that we need to create a customer class and every new method or path will require explicit configuration.

Author: changed to this:
`The downside is that we need to create custom classes and with every new method or path we want to serve, we have to add an explicit configuration. `

> Also, we need to understand API Gateway request event structure in order to decapsulate requested path and http method. 

Reviewer: Hmm, can you elaborate on why it is a downside, or what problem it might cause? I assume you mean we need to know the directory structure of the event object (where we did not have error handling). It might need more clarification.

Author: I removed this part, after a second consideration, it makes no sense :-). I rewrote this section to this: 
`This approach allows us to simplify the configuration of our infrastructure since we have added all Gateway API paths in the `HelloWorldFunction` event section. We need to understand the internal structure of the API Gateway request events though, to deduce the requested path and the http method. This requires us to put additional engineering effort to provide proper error handling. Also, if we decide to use another event source for our lambda, since we are highly coupled it would require rewriting of our lambda handler to get the information we need.`
 Do you think it is accurate?

> In order to mitigate this downsides we use Lambda Powertools.

Reviewer: Convert to positive: Lets see how we can improve it with Powertools

Author: Converted

=== "app.py"

    ```python hl_lines="1 4 7 12 18"
    from aws_lambda_powertools.event_handler.api_gateway import ApiGatewayResolver

    app = ApiGatewayResolver()

    @app.get("/hello/<name>")
    def hello_name(name):
        return {"statusCode": 200, "body": {"message": f"hello {name}!"}}

    @app.get("/hello")
    def hello():
        return {"statusCode": 200, "body": {"message": "hello unknown!"}}

    def lambda_handler(event, context):
        return app.resolve(event, context)
    ```
=== "requirements.txt"

    ```bash
    aws-lambda-powertools
    ```

> Since we used powertools api_gateway event handler we don't need to know internal event structure. All we do we instantiate `ApiGatewayResolver` class and we call it in our `lambda_handler`.

Reviewer: Convert to positive: Powertools provides a class `ApiGatewayResolver` that helps you understand the structure, no need to look it up.

Author: Converted as suggested:
`Powertools provides an `ApiGatewayResolver` class, which helps you understand the structure, no need to look it up.
Because we have added the route annotation as a decorator for our methods, we may directly use the parameters passed in the query.`

> Since we added route annotation as a decorator for our methods we can directly use parameters passed in request.

Reviewer: 2nd sentence starting with "Since we..." you can reverse the order of the logic here and start with the value, instead of "we get foo by doing bar" you can say: "We can access parameters, by using route annotation".

Author: Rewrote to:
`Powertools provides an `ApiGatewayResolver` class, which helps you understand the structure, no need to look it up.
We can also directly use the parameters passed in the request now, because we have added the route annotation as the decorator for our methods.
For SAM to build our package correctly, we have specified lambda powertools package in our `requirement.txt` file. `

> In order for SAM to build our package correctly we specified lambda powertools package in our requirements file. 

Reviewer: This is prerequisite, move it higher to the beginning, otherwise we already know it.

Author: Not sure about this, since we follow a more narrative approach in the quickstart powertools is not required to start. In first stage we don't use powertools we get there after we try multiple lambdas approach and custom Router.

> It is then automatically downloaded by SAM and added to lambda zip file. 

Reviewer: can be removed, or pulled higher to "prerequisites" that `sam build` pull the dependencies.

!!! tip
    If you are interested in learning how python decorators work under the hood you can follow great article from [Real Python](https://realpython.com/primer-on-python-decorators/).
## Structured Logging

> In next step you have been tasked to propose production grade logging capabilities to your lambda code.

Reviewer: Maybe convert to "We want to have production grade...", to have more guiding character in the quickstart 🤔

Author: Not sure whad do you mean by this. Can you elaborate?

You decided to follow [structured logging approach](https://docs.aws.amazon.com/lambda/latest/operatorguide/parse-logs.html) which means that you don’t write hard-to-parse and hard-to-keep-consistent prose in your logs but that you log events that happen in a context instead. 

> We want our log event to be in JSON format. This will allow us to search through it easily in CloudWatch Logs or Cloudwatch Insight.

Reviewer: This is the key information presented in the 3rd sentence, let's pull it higher and then explain why it is important: hard-to-prase, link for more resources etc.

Author: Refactored to:
`In the next step, you have been given the task of proposing production quality logging capabilities to your lambda code.
We want our log event to be in a JSON format. Also, You decided to follow [structured logging approach](https://docs.aws.amazon.com/lambda/latest/operatorguide/parse-logs.html). In a result, we expect easy to search, consistent logs containing enough context and data to analyse the status of our system. We can take advantage of CloudWatch Logs and Cloudwatch Insight for this purpose.`
Let me know what do you think

> Initial idea might be to initiate logger from standard library and use `pythonjsonlogger` library to simplify structured log creation.

Reviewer: We have the word **Initial** twice, also simplify: python logger is our first choice for simple structured logging.

Author: Refactored to:
`The first option could be to use a python logger in combination with the `pythonjsonlogger` library for simple structured logging.
`

=== "app.py"

    ```python hl_lines="1 4 8-13 20 26 31"
    import logging
    import os

    from pythonjsonlogger import jsonlogger

    from aws_lambda_powertools.event_handler.api_gateway import ApiGatewayResolver

    logger = logging.getLogger("APP")
    logHandler = logging.StreamHandler()
    formatter = jsonlogger.JsonFormatter(fmt="%(asctime)s %(levelname)s %(name)s %(message)s")
    logHandler.setFormatter(formatter)
    logger.addHandler(logHandler)
    logger.setLevel(os.getenv("LOG_LEVEL", "INFO"))

    app = ApiGatewayResolver()

    @app.get("/hello/<name>")
    def hello_name(name):
        logger.info(f"Request from {name} received")
        return {"statusCode": 200, "body": {"message": f"hello {name}!"}}

    @app.get("/hello")
    def hello():
        logger.info(f"Request from unknown received")
        return {"statusCode": 200, "body": {"message": "hello unknown!"}}

    def lambda_handler(event, context):
        logger.debug(event)
        return app.resolve(event, context)
    ```
=== "requirements.txt"

    ```bash
    aws-lambda-powertools
    python-json-logger
    ```

> So in our first attempt we create new APP Logger. We configure Log Handler, formatter and log level. Then we use this logger to log required information. We will see logs structured like that

Reviewer: If you want to highlight the complexity because of number of steps, you can have an enumeration: 
* create  app logger
* configure handler and formatter
* set log level

Just a thought, it doesn't have to be a list, can also be in one sentence.

Author: Enumeration looks good, fixing

```json
{"asctime": "2021-11-22 15:32:02,145", "levelname": "INFO", "name": "APP", "message": "Request from unknown received"}
```
instead of
```json
[INFO]  2021-11-22T15:32:02.145Z        ba3bea3d-fe3a-45db-a2ce-72e813d55b91    Request from unknown received
```

> So far so good! You might also decide to add additional context into your logs.

Reviewer: You  might decide -> We want to have

Author: fixed

> You can take it from lambda context or event. Downside of this approach is that we would add additional complexity to ensure that you always add those specific attributes in every place.

Reviewer: This is not clear, how to do that "from lambda context or event" and it's not clear what "additional complexity" we will have. Maybe we need an example here?

Author: Remove additional complexity and rewrote it:
`
So far, so good! To make things easier, we want to add extra context to the logs. We can extract it from a lambda context or an event passed to lambda handler at the time of invocation. We have to make sure that we always add those specific attributes wherever a logger is used. Can we ensure that the required attributes are added automatically on our behalf without having to move them around? Yes! Powertools Logger to the rescue :-)
~

Do you think it is enough?

> Also, we added boilerplate code and even with additional third party library it is not straightforward. 

Reviewer: Hypothesis without example, maybe leave it out and focus that powertools make logging easier.

Author: Removed that part and focused on benefit of using powertools:
`

So far, so good! To make things easier, we want to add extra context to the logs. We can extract it from lambda context or event. We have to make sure that we always add those specific attributes wherever a logger is used. Can we ensure that the required attributes are added automatically on our behalf without having to move them around? Yes! Powertools Logger to the rescue :-)
`

Can we do better? Yes! We can use powertools logger capabilities as with next example.
=== "app.py"

    ```python hl_lines="1 6 13 19 23"
    from aws_lambda_powertools import Logger
    from aws_lambda_powertools.logging import correlation_paths

    from aws_lambda_powertools.event_handler.api_gateway import ApiGatewayResolver

    logger = Logger(service="APP")

    app = ApiGatewayResolver()

    @app.get("/hello/<name>")
    def hello_name(name):
        logger.info(f"Request from {name} received")
        return {"statusCode": 200, "body": {"message": f"hello {name}!"}}

    @app.get("/hello")
    def hello():
        logger.info(f"Request from unknown received")
        return {"statusCode": 200, "body": {"message": "hello unknown!"}}

    @logger.inject_lambda_context(correlation_id_path=correlation_paths.API_GATEWAY_REST, log_event=True)
    def lambda_handler(event, context):
        return app.resolve(event, context)
    ```

> We imported powertools Logger class which we created instance from in line 6. And this is basically it! By using powertools logger we have all required configuration done out of the box. 

Reviewer: Can be simplified into one sentence: We add powertools logger in line 6 and all the configuration is done.

Author: changed

> We used `logger.inject_lambda_context` decorator to inject lambda context into every log. We also instruct logger to log correlation id taken from API Gateway and to log coming event automatically. 

Reviewer: simplify: and ~~to log~~ events automatically.

Author: done

> Correlation id will help with tracking request between api gateway and lambda.

Reviewer: How does it help?

Author: Rephrased to:
`Because powertools library adds a correlation identifier to each log, we can easily correlate all the logs generated for a specific request.`

As a result, we should see logs with following attributes.
=== "Example Application Structured Log"
```json
{
    "level":"INFO",
    "location":"hello:17",
    "message":"Request from unknown received",
    "timestamp":"2021-10-22 16:29:58,367+0000",
    "service":"hello",
    "sampling_rate":"0.1",
    "cold_start":true,
    "function_name":"HelloWorldFunction",
    "function_memory_size":"256",
    "function_arn":"arn:aws:lambda:us-east-1:123456789012:function:HelloWorldFunction",
    "function_request_id":"d50bb07a-7712-4b2d-9f5d-c837302221a2",
    "correlation_id":"bf9b584c-e5d9-4ad5-af3d-db953f2b10dc"
    }
```
By having structured logs like this, we might easily search and analyse them in [CloudWatch Logs Insight](https://docs.aws.amazon.com/AmazonCloudWatch/latest/logs/AnalyzingLogData.html).
=== "CloudWatch Logs Insight Example"
![CloudWatch Logs Insight Example](./media/cloudwatch_logs_insight_example.png)
## Tracing

> Next task you decided to work on is to add proper tracing mechanism to your stack. Developers want to be able to analyse request traces coming through API Gateway to your Lambda. 

Reviewer: You -> We

Author: done

Together with structured logs it will be a great step to providing observability for your application!

> Ideal service for this purpose is [AWS X-RAY](https://aws.amazon.com/xray/). How can we send traces from our application to the AWS X-RAY service?

Reviewer: Reduce marketing, ideal is a strong word 😃

Author: True :-) Rephrased to:
`The AWS service that has these capabilities is [AWS X-RAY](https://aws.amazon.com/xray/). How do we send application trace to the AWS X-RAY service then?
Let's first explore how we can achieve this with [x-ray SDK](https://docs.aws.amazon.com/xray-sdk-for-python/latest/reference/index.html), and then try to simplify it with the Powertools library.`

                return app.resolve(event, context)
    ```

=== "template.yaml"

    ```yaml hl_lines="15"
    AWSTemplateFormatVersion: "2010-09-09"
    Transform: AWS::Serverless-2016-10-31
    Description: >
        Sample SAM Template for powertools-quickstart
    Globals:
    Function:
        Timeout: 3
    Resources:
        HelloWorldFunction:
            Type: AWS::Serverless::Function
            Properties:
                CodeUri: hello_world/
                Handler: app.lambda_handler
                Runtime: python3.9
                Tracing: Active
                Events:
                    HelloWorld:
                    Type: Api
                    Properties:
                        Path: /hello
                        Method: get
                    HelloWorldName:
                    Type: Api 
                    Properties:
                        Path: /hello/{name}
                        Method: get
    Outputs:
        HelloWorldApi:
            Description: "API Gateway endpoint URL for Prod stage for Hello World function"
            Value: !Sub "https://${ServerlessRestApi}.execute-api.${AWS::Region}.amazonaws.com/Prod/hello/"
    ```

> Few things happened here. Firstly, we imported required x-ray SDK classes. 

Reviewer: First~~ly~~

Author: fixed

> `xray_recorder` is instance of global AWS X-Ray recorder class that will begin/end segments/subsegments and send them to the X-Ray daemon.

Reviewer: is **an** instance of **a** global

Author: rephrased to:
``A lot has happened here. First, we imported required X-ray SDK classes. 
`xray_recorder` is a global AWS X-ray recorder class instance that starts/ends segments/sub-segments and sends them to the X-ray daemon.`

> In order to create new sub-segments we use `xray_recorder.in_subsegment` method as context manager. 

Reviewer: Simplify:  ~~In order~~ to create a new

Author: Rewrote to:
To build new sub-segments, we use `xray_recorder.in_subsegment` method as a context manager. `

By using custom subsegments we were able to add granular sub-traces visible in our X-RAY console, separated by called method name.

> Also we track lambda cold start by setting global variable outside of handler. Variable is set only during lambda runtime initialization. 

Reviewer: We also track.... 
The variable is set... -> provide benefit here, it's not clear why it's important (cache)

Author: Changed to:
`Also, we track lambda cold start by setting global variable outside of a handler. The variable is defined only upon lambda initialization. This information provides an overview of how often the runtime is reused by lambda invoked, which directly impacts perceived lambda performance and latency.`

> In order to enable tracing for our Lambda we need to configure it in our SAM template.yaml and add `Tracing: Active` under lambda `Properties` section.

Reviewer: Simplify: ~~In order~~ to

Author: Rewrote to:
`To allow the tracking of our Lambda, we need to set it up in our SAM template and add `Tracing: Active` under lambda `Properties` section.`

!!! Info 
    Interested in diving deep in Context Managers and understanding benefits of using them? Follow [article](https://realpython.com/python-with-statement/) from Real Python.
!!! Info

>    If you want to understand how Lambda execution environment works and why cold starts may occur, follow great [blog series](https://aws.amazon.com/blogs/compute/operating-lambda-performance-optimization-part-1/.

Reviewer: follow **a**/**this** great

Author: Rewrite sentence to:
`If you want to understand how the Lambda execution environment works and why cold starts can occur, follow [blog series](https://aws.amazon.com/blogs/compute/operating-lambda-performance-optimization-part-1/.`

=== "app.py"

    ```python hl_lines="1 11 13 19 21 27"
    from aws_lambda_powertools import Logger, Tracer
    from aws_lambda_powertools.event_handler.api_gateway import ApiGatewayResolver
    from aws_lambda_powertools.logging import correlation_paths

    logger = Logger(service="APP")
    tracer = Tracer()
    app = ApiGatewayResolver()

    @app.get("/hello/<name>")
    @tracer.capture_method
    def hello_name(name):
        tracer.put_annotation("User", name)
        logger.info(f"Request from {name} received")
        return {"statusCode": 200, "body": {"message": f"hello {name}!"}}

    @app.get("/hello")
    @tracer.capture_method
    def hello():
        tracer.put_annotation("User", "unknown")
        logger.info(f"Request from unknown received")
        return {"statusCode": 200, "body": {"message": "hello unknown!"}}

    @logger.inject_lambda_context(correlation_id_path=correlation_paths.API_GATEWAY_REST, log_event=True)
    @tracer.capture_lambda_handler
    def lambda_handler(event, context):
        return app.resolve(event, context)
    ```
    
Thanks to using powertools tracer we have much cleaner code right now. 

> In order to make our methods visible in traces we added `@tracer.capture_method` decorator to processing methods. 

Reviewer: ~~In order~~

Author: Rewrote to:
`Thanks to the help of powertools tracer we have much cleaner code right now. 
To make our methods visible in the traces, we added @tracer. capture_method' decorator to the processing methods.`

> We added  annotations directly in code without adding it with context manager using `tracer.put_annotation` method. 

Reviewer: two space characters after "We added "

Author: Fixed

It also automatically adds lambda response as a metadata into trace so we don't need to worry about it as well.
!!! tip 
    To read about differences between annotation and metadata in traces please follow [link](https://awslabs.github.io/aws-lambda-powertools-python/latest/core/tracer/#annotations-metadata). 
As a result, you should see traces for your lambda in the X-RAY console.
=== "Example X-RAY Console View"
![Tracer utility](./media/tracer_utility_showcase_2.png)

You may also consider using **CloudWatch ServiceLens** which ties together CloudWatch metrics and logs, in addition to traces from AWS X-Ray.
It gives you a complete view of your applications and their dependencies enhancing the observability of your services. 
From there you can easily navigate to specific logs in CloudWatch Logs Insight, Metrics Dashboard or Traces in CloudWatch X-Ray traces.
=== "Example CloudWatch ServiceLens View"
![CloudWatch ServiceLens View](./media/tracer_utility_showcase_3.png)
!!! Info
    If you want to learn more about CloudWatch ServiceLens follow this [link](https://docs.aws.amazon.com/AmazonCloudWatch/latest/monitoring/ServiceLens.html)
## Custom Metrics

> Last step in order to provide full observability is to add some business level metrics. 

Reviewer: Some comma is missing here or we might need to restructure the sentence:

In the last step we want to have more observability by adding business level metrics.

Author: Changed to
`The final step to provide complete observability is to add certain measures at the business level. `

Lambda adds low level technical metrics (Invocations, Duration, Error Count & Success Rate etc) into CloudWatch Metrics out of the box. 

> You have decided to add some custom metrics in order to increase visibility into what happens with your application. 

Reviewer: Can be removed, since it repeats the information for the first sentence.

Author: done

    Resources:
    HelloWorldFunction:
        Type: AWS::Serverless::Function
        Properties:
        CodeUri: hello_world/
        Handler: app.lambda_handler
        Runtime: python3.9
        Architectures:
            - x86_64
        Events:
            HelloWorld:
            Type: Api 
            Properties:
                Path: /hello
                Method: get
    Outputs:
        HelloWorldApi:
            Description: "API Gateway endpoint URL for Prod stage for Hello World function"
            Value: !Sub "https://${ServerlessRestApi}.execute-api.${AWS::Region}.amazonaws.com/Prod/hello/"
    ```
Our lambda code consists of the `lambda_handler` method that is invoked by the API and the `hello` method that returns the results to the API gateway and is invoked by the handler itself.
The SAM model configures API Gateway, which redirects traffic to Lambda for one path only: `hello`.
!!! Warning 
    For simplicity, we do not set up authentication and authorisation in the example!
### Run your code
At each point, you have two ways to run your code. Locally and within your AWS account. Given that we use SAM, the two methods are just as simple.
#### Local test

> SAM allows you to execute a server-less application locally. Perform the next command in a shell.

Reviewer: serverless instead of server-less?

Author: fixed

    Transform: AWS::Serverless-2016-10-31
    Description: > 
        Sample SAM Template for powertools-quickstart
    Globals:
        Function:
            Timeout: 3
    Resources:
        HelloWorldFunction:
            Type: AWS::Serverless::Function
            Properties:
            CodeUri: hello_world/
            Handler: app.lambda_handler
            Runtime: python3.9
            Events:
                HelloWorld:
                Type: Api 
                Properties:
                    Path: /hello
                    Method: get
        HelloWorldFunctionName:
            Type: AWS::Serverless::Function
            Properties:
            CodeUri: hello_world/
            Handler: app_name.lambda_handler
            Runtime: python3.9
            Events:
                HelloWorldName:
                Type: Api
                Properties:
                    Path: /hello/{name}
                    Method: get
    Outputs:
        HelloWorldApi:
            Description: "API Gateway endpoint URL for Prod stage for Hello World function"
            Value: !Sub "https://${ServerlessRestApi}.execute-api.${AWS::Region}.amazonaws.com/Prod/hello/"
    ```

> This way certainly works for simple use-case. But what happens if your application gets bigger and we need to cover numerous URL paths and HTTP methods for them? If that is the case, we should:

Reviewer: use case instead of use-case

Author: fixed

    ```

=== "template.yaml"

    ```yaml hl_lines="15-25"
        AWSTemplateFormatVersion: "2010-09-09"
        Transform: AWS::Serverless-2016-10-31
        Description: >
            Sample SAM Template for powertools-quickstart
        Globals:
            Function:
            Timeout: 3
        Resources:
            HelloWorldFunction:
            Type: AWS::Serverless::Function
            Properties:
                CodeUri: hello_world/
                Handler: app.lambda_handler
                Runtime: python3.9
                Events:
                    HelloWorld:
                        Type: Api
                        Properties:
                        Path: /hello
                        Method: get
                    HelloWorldName:
                        Type: Api 
                        Properties:
                        Path: /hello/{name}
                        Method: get
        Outputs:
            HelloWorldApi:
                Description: "API Gateway endpoint URL for Prod stage for Hello World function"
                Value: !Sub "https://${ServerlessRestApi}.execute-api.${AWS::Region}.amazonaws.com/Prod/hello/"
    ```

> * We added two `hello_name` and `hello` methods (line 1-7). 

Reviewer: either remove the word two or bring the word method to the front:
- We added two methods: `hello_name` and `hello`
- We added `hello_name` and `hello` methods.

Author: fixed

* We added the `Router` class which allows us to record the method that should be called when the specific request arrives (line 10-23). 
* We created the instance and added the configuration with the mapping of the processing methods and the http query method (line 26-28). 
* In the lambda handler, we call router instance `get` method to retrieve a reference to the processing method (`hello` or `hello_name`) that will process the query (line 34). 
* Finally, we run this method and send the results back to API Gateway (line 35).

> This approach allows us to simplify the configuration of our infrastructure since we have added all Gateway API paths in the `HelloWorldFunction` event section. We need to understand the internal structure of the API Gateway request events though, to deduce the requested path, http method and path parameters. This requires us to put additional engineering effort to provide proper error handling. Also, if we decide to use another event source for our lambda, since we are highly coupled it would require rewriting of our lambda handler to get the information we need.

Reviewer: Gateway API -> API Gateway

Author: fixed

    app = ApiGatewayResolver()

    @app.get("/hello/<name>")
    def hello_name(name):
        logger.info(f"Request from {name} received")
        return {"statusCode": 200, "body": {"message": f"hello {name}!"}}

    @app.get("/hello")
    def hello():
        logger.info(f"Request from unknown received")
        return {"statusCode": 200, "body": {"message": "hello unknown!"}}

    def lambda_handler(event, context):
        logger.debug(event)
        return app.resolve(event, context)
    ```
=== "requirements.txt"

    ```bash
    aws-lambda-powertools
    python-json-logger
    ```
On the first try, we did a couple of steps to set up our logging:

* Create an application logger called `APP`.
* Configure handler and formatter.
* Set log level.

After that, we use this logger in our application code to record the required information. We see logs structured as follows:
```json
{"asctime": "2021-11-22 15:32:02,145", "levelname": "INFO", "name": "APP", "message": "Request from unknown received"}
```
instead of
```json
[INFO]  2021-11-22T15:32:02.145Z        ba3bea3d-fe3a-45db-a2ce-72e813d55b91    Request from unknown received
```

> So far, so good! To make things easier, we want to add extra context to the logs. We can extract it from a lambda context or an event passed to lambda handler at the time of invocation. We have to make sure that we always add those specific attributes wherever a logger is used. Can we ensure that the required attributes are added automatically on our behalf without having to move them around? Yes! Powertools Logger to the rescue :-)

Reviewer: We have to make sure that we always -> We need

Author: fixed

As a result, we should see logs with following attributes.
=== "Example Application Structured Log"
```json
{
    "level":"INFO",
    "location":"hello:17",
    "message":"Request from unknown received",
    "timestamp":"2021-10-22 16:29:58,367+0000",
    "service":"hello",
    "sampling_rate":"0.1",
    "cold_start":true,
    "function_name":"HelloWorldFunction",
    "function_memory_size":"256",
    "function_arn":"arn:aws:lambda:us-east-1:123456789012:function:HelloWorldFunction",
    "function_request_id":"d50bb07a-7712-4b2d-9f5d-c837302221a2",
    "correlation_id":"bf9b584c-e5d9-4ad5-af3d-db953f2b10dc"
    }
```
By having structured logs like this, we might easily search and analyse them in [CloudWatch Logs Insight](https://docs.aws.amazon.com/AmazonCloudWatch/latest/logs/AnalyzingLogData.html).
=== "CloudWatch Logs Insight Example"
![CloudWatch Logs Insight Example](./media/cloudwatch_logs_insight_example.png)
## Tracing

> The next task we have chosen to work on is to add an appropriate tracking mechanism to your stack. Developers want to be able to analyze traces of queries that pass via the API gateway to your Lambda. 

Reviewer: want to be able to -> want

Author: fixed

        Sample SAM Template for powertools-quickstart
    Globals:
    Function:
        Timeout: 3
    Resources:
        HelloWorldFunction:
            Type: AWS::Serverless::Function
            Properties:
                CodeUri: hello_world/
                Handler: app.lambda_handler
                Runtime: python3.9
                Tracing: Active
                Events:
                    HelloWorld:
                    Type: Api
                    Properties:
                        Path: /hello
                        Method: get
                    HelloWorldName:
                    Type: Api 
                    Properties:
                        Path: /hello/{name}
                        Method: get
    Outputs:
        HelloWorldApi:
            Description: "API Gateway endpoint URL for Prod stage for Hello World function"
            Value: !Sub "https://${ServerlessRestApi}.execute-api.${AWS::Region}.amazonaws.com/Prod/hello/"
    ```

A lot has happened here. First, we imported required X-ray SDK classes. 
`xray_recorder` is a global AWS X-ray recorder class instance that starts/ends segments/sub-segments and sends them to the X-ray daemon.

> To build new sub-segments, we use `xray_recorder.in_subsegment` method as a context manager. 

Reviewer: sub-segments -> subsegments

> Using customized sub-segments, we were able to add visible granular sub-traces into our X-ray, separated by the method name.

Author: done

> Also, we track lambda cold start by setting global variable outside of a handler. The variable is defined only upon lambda initialization. This information provides an overview of how often the runtime is reused by lambda invoked, which directly impacts perceived lambda performance and latency.

Reviewer: ~~perceived~~

Author: done

=== "app.py"

    ```python hl_lines="1 11 13 19 21 27"
    from aws_lambda_powertools import Logger, Tracer
    from aws_lambda_powertools.event_handler.api_gateway import ApiGatewayResolver
    from aws_lambda_powertools.logging import correlation_paths

    logger = Logger(service="APP")
    tracer = Tracer()
    app = ApiGatewayResolver()

    @app.get("/hello/<name>")
    @tracer.capture_method
    def hello_name(name):
        tracer.put_annotation("User", name)
        logger.info(f"Request from {name} received")
        return {"statusCode": 200, "body": {"message": f"hello {name}!"}}

    @app.get("/hello")
    @tracer.capture_method
    def hello():
        tracer.put_annotation("User", "unknown")
        logger.info(f"Request from unknown received")
        return {"statusCode": 200, "body": {"message": "hello unknown!"}}

    @logger.inject_lambda_context(correlation_id_path=correlation_paths.API_GATEWAY_REST, log_event=True)
    @tracer.capture_lambda_handler
    def lambda_handler(event, context):
        return app.resolve(event, context)
    ```
    

> Thanks to the help of powertools tracer we have much cleaner code right now. 

Reviewer: Thanks to the help of powertools -> With powertools

Author: done

It also automatically adds lambda response as a metadata into trace, so we don't need to worry about it as well.
!!! tip 
    For differences between annotations and metadata in traces, please follow [link](https://awslabs.github.io/aws-lambda-powertools-python/latest/core/tracer/#annotations-metadata). 
Therefore, you should see traces of your lambda in the X-ray console.
=== "Example X-RAY Console View"
![Tracer utility](./media/tracer_utility_showcase_2.png)

You may also consider using **CloudWatch ServiceLens** which links the CloudWatch metrics and logs, in addition to traces from the AWS X-Ray.
It gives you a complete view of your apps and their dependencies, making your services more observable. 
From here, you can easily browse to specific logs in CloudWatch Logs Insight, Metrics Dashboard or Traces in CloudWatch X-Ray traces.
=== "Example CloudWatch ServiceLens View"
![CloudWatch ServiceLens View](./media/tracer_utility_showcase_3.png)
!!! Info
    For more information on CloudWatch ServiceLens, please visit [link](https://docs.aws.amazon.com/AmazonCloudWatch/latest/monitoring/ServiceLens.html).
## Custom Metrics

> The final step to provide complete observability is to add certain measures at the business level. 

Reviewer: certain measures -> metrics (or business metrics). You can also drop one or two examples such as number of sales or reservations.

Author: Added `The final step to provide complete observability is to add business metrics (such as number of sales or reservations).`

> Lambda adds low-level technical metrics (such as Invocations, Duration, Error Count & Success Rate) to the CloudWatch metrics out of the box. 

Reviewer: low-level -> low level

Author: removed low level

        return app.resolve(event, context)
    ```
=== "template.yaml"

    ```yaml hl_lines="27 28"
    AWSTemplateFormatVersion: "2010-09-09"
    Transform: AWS::Serverless-2016-10-31
    Description: >
        Sample SAM Template for powertools-quickstart
    Globals:
        Function:
            Timeout: 3
    Resources:
        HelloWorldFunction:
            Type: AWS::Serverless::Function
            Properties:
            CodeUri: hello_world/
            Handler: app.lambda_handler
            Runtime: python3.9
            Tracing: Active
            Events:
                HelloWorld:
                Type: Api
                Properties:
                    Path: /hello
                    Method: get
                HelloWorldName:
                Type: Api
                Properties:
                    Path: /hello/{name}
                    Method: get
            Policies:
                - CloudWatchPutMetricPolicy: {}
    Outputs:
        HelloWorldApi:
            Description: "API Gateway endpoint URL for Prod stage for Hello World function"
            Value: !Sub "https://${ServerlessRestApi}.execute-api.${AWS::Region}.amazonaws.com/Prod/hello/"

    ```

> To add our custom metric in **CloudWatch** we added `boto3` cloudwatch client followed by creation of the new `put_metric_data` method that uses this client to put the metric in CloudWatch synchronously. We call it in our method `hello` and `hello_name`. We divide our measures by type of application and method. Thus, we can follow the frequency at which specific methods are called. We also need to add additional inline policy allowing our lambda to write metrics in the CloudWatch. In `template.yaml` we added `CloudWatchPutMetricPolicy` policy.

Reviewer: measures -> metrics

Author: done

---
title: QuickStart
description: Powertools introduction
---
Quickstart introducing core Powertools functionalities.

> 

Author: done

> 

Author: Added requirements heading

> ## Installation 

Reviewer 2: As you're covering **1/** Cloning the project, **2/** Making small modifications, **3/** Running it locally, you're past the scope of an Installation header -- **Getting started** makes this more explicit and gives you room for additional changes you might want to make.

```suggestion
## Getting started 
```

Author: Changed as suggested

> With [SAM](https://docs.aws.amazon.com/serverless-application-model/latest/developerguide/serverless-sam-cli-install.html) installed, you can follow this Quickstart step by step or you can build a project with the final Quickstart code you can play with.
> If you follow it all along, create a new empty project.

Reviewer 2: Depending on the their primary spoken language this can be confusing (too many paths to take). What you can do is be direct upfront on what's the main path, and what's the alternative one.

```suggestion
With [SAM](https://docs.aws.amazon.com/serverless-application-model/latest/developerguide/serverless-sam-cli-install.html) installed, let's clone our sample project before we add one feature at a time.

???+ tip "Tip: Want to skip to the final project?"
    Bootstrap directly via SAM CLI: `sam init --location https://github.com/aws-samples/cookiecutter-aws-sam-python`
```

Author: good one! Will follow this approach. Thx!

=== "shell"
```bash

> sam init --runtime python3.9 --dependency-manager pip --app-template hello-world --name powertools-quickstart

Reviewer 2: Great call out in using long parameter names to inform what each flag does!

Author: explicit is better than implicit :-) [ Zen of Python](https://towardsdatascience.com/the-zen-of-python-a-guide-to-pythons-design-principles-93f3f76d088a)

```
If you want to play with the final code, then you can download it from [aws-samples](https://github.com/aws-samples/cookiecutter-aws-sam-python) github repository.
=== "shell"
```bash
sam init --location https://github.com/aws-samples/cookiecutter-aws-sam-python
```
### Configuration

> If you deploy code from Quickstart make sure to [set up your AWS credentials](https://docs.aws.amazon.com/serverless-application-model/latest/developerguide/serverless-getting-started-set-up-credentials.html).

Reviewer 2: Call this out under Requirements, not here. It might be too late already. If they already have, it's one less section to scroll

Author: Moved to requirement section

> ### Code example

Reviewer 2: If it's not too much to ask, can we have a diagram of the sample app we're doing (API GW -> Lambda)?

A basic image can help people new to this visualize what it is that they're doing.

Author: Sure can draw it, what is the standard tooling you use for diagram drawing in our documentation? Where is the best place to put it, what section?

Reviewer 2: We gravitate towards drawio to make it easier to extend later.

> In the `powertools-quickstart` folder, we modify the following files:

Reviewer 2: ```suggestion
Within the `powertools-quickstart` folder, we will modify the following files:
```

* **app.py** - Application code.
* **template.yaml** - AWS infrastructure configuration using SAM.

> * **requirements.txt** - List of extra Python packages needed.

Reviewer 2: Was this perhaps a leftover? This is not being changed. I suspect you had it at some point, then used Lambda Layers, then removed, etc.

Paths you can take:

1. Include Lambda Powertools Layer (preferred) in the template
2. Include the change necessary within requirements.txt

In either option, they will have to create a virtual env & install Lambda Powertools as a development dependency in order to gain auto-complete / IntelliSense in their IDE.

Author: We modify requirement.txt file in *API Gateway router* section when we introduce powertools. My intention here was to give a summary what files we change as we progress with our journey. Will make it more explicit to avoid confusion

Author: Modified this piece of documentation to this:
```
### Project structure
As we move forward, we will modify the following files within the `powertools-quickstart` folder:

* **app.py** - Application code.
* **template.yaml** - AWS infrastructure configuration using SAM.
* **requirements.txt** - List of extra Python packages needed.
```
Let me know if this is ok

Reviewer 2: Looks awesome! thank you!

Let's configure our base application to look like the following code snippet.
=== "app.py"

    ```python
    def hello():

>         return {"statusCode": 200, "body": {"message": "hello unknown!"}}

Reviewer 2: shouldn't the value of `body` be a JSON string?

Author: In this particular example I guess it doesn't matter really, but in general I agree. By default lambda returns `Content-Type: application/json` header so the response should be formatted accordingly. I will change the code to utilise `json.dumps` to be consistent

Author: Modifed all examples to return json

    def lambda_handler(event, context):
        return hello()
    ```

> === "template.yaml"
> 
>     ```yaml
>     AWSTemplateFormatVersion: '2010-09-09'
>     Transform: AWS::Serverless-2016-10-31
>     Description: >
>         Sample SAM Template for powertools-quickstart
>     Globals:
>     Function:
>         Timeout: 3
>     Resources:
>     HelloWorldFunction:
>         Type: AWS::Serverless::Function
>         Properties:
>         CodeUri: hello_world/
>         Handler: app.lambda_handler
>         Runtime: python3.9
>         Architectures:
>             - x86_64
>         Events:
>             HelloWorld:
>             Type: Api 
>             Properties:
>                 Path: /hello
>                 Method: get
>     Outputs:
>         HelloWorldApi:
>             Description: "API Gateway endpoint URL for Prod stage for Hello World function"
>             Value: !Sub "https://${ServerlessRestApi}.execute-api.${AWS::Region}.amazonaws.com/Prod/hello/"
>     ```

Reviewer 2: Your indentation is off and will accidentally fail deployment - it might have been any markdown plugin you have to auto-format it (happened to me before)

```suggestion
=== "template.yaml"

    ```yaml
    AWSTemplateFormatVersion: '2010-09-09'
    Transform: AWS::Serverless-2016-10-31
    Description: Sample SAM Template for powertools-quickstart
    Globals:
        Function:
            Timeout: 3
    Resources:
        HelloWorldFunction:
            Type: AWS::Serverless::Function
            Properties:
                CodeUri: hello_world/
                Handler: app.lambda_handler
                Runtime: python3.9
                Architectures:
                    - x86_64
                Events:
                    HelloWorld:
                      Type: Api 
                      Properties:
                          Path: /hello
                          Method: get
    Outputs:
        HelloWorldApi:
            Description: "API Gateway endpoint URL for Prod stage for Hello World function"
            Value: !Sub "https://${ServerlessRestApi}.execute-api.${AWS::Region}.amazonaws.com/Prod/hello/"
    ```
```

Author: Thanks, fixed all templates to have correct indetation

> Our lambda code consists of the `lambda_handler` method that is invoked by the API and the `hello` method that returns the results to the API gateway and is invoked by the handler itself.

Reviewer 2: Tip: Try keeping a sentence under 20 words or less - doesn't always work. This helps readers break down the information into smaller bits, process, and carry on as if they were reading a story.

```suggestion
Our Lambda code consists of an entry point function named `lambda_handler`, and a `hello` function. 

When API Gateway receives a request, Lambda will call our `lambda_handler` function, subsequently calling the `hello` function. API Gateway will use this response to return the correct HTTP Status Code and payload back to the caller.
```

The way I personally do these days as screen width matters is: **1/** Read the sentence out and loud and see if you get the main idea upfront -- too many paths/ideas means you need to break it down, **2/** Run the docs locally to see the aesthetic aspect of it, so it doesn't become too jammed.

Author: Thanks for the great tip, your sentence make much more sense

The SAM model configures API Gateway, which redirects traffic to Lambda for one path only: `hello`.
!!! Warning 

>     For simplicity, we do not set up authentication and authorisation in the example!

Reviewer 2: Tip: Convert to an actionable suggestion in case they do want to pursue this -- As a customer, where would I go to learn more on doing this?

Author: Extended Warning message with link to appropiate documentation.

### Run your code
At each point, you have two ways to run your code. Locally and within your AWS account. Given that we use SAM, the two methods are just as simple.
#### Local test
SAM allows you to execute a serverless application locally. Perform the next command in a shell.
```bash
> sam build && sam local start-api
...
2021-11-26 17:43:08  * Running on http://127.0.0.1:3000/ (Press CTRL+C to quit)
```

> As a result API endpoint will be exposed for you. You can trigger it with the 'curl' command like in the following example.

Reviewer 2: Tip: Not everyone might have curl installed (think Windows customers, unless they use WSL). Mention other forms like browser, or REST API clients.

Author: Added another tip:
```
???+ tip "Tip: Want to use different tool for API testing?"
    Instead of `curl` you have many other options to choose from.

    * Choose different command line tool like `HTTPie`.
    * Type the url in the browser directly
    * Use REST API client like `Postman` or `SoupUI`.
```

```bash
> curl http://127.0.0.1:3000/hello
{"statusCode":200,"body":{"message":"hello unknown!"}}%
```
!!! info
    To learn more about local testing, please visit [SAM local testing](https://docs.aws.amazon.com/serverless-application-model/latest/developerguide/sam-cli-command-reference-sam-local-start-api.html) page
!!! warning

>     **Powertools Tracer** requires X-RAY service to work. This means that you will not see the traces locally. Roll it out on your AWS account instead.

Reviewer 2: QuestionL **Is this the right place?**, If you want to preemptively inform about local traces, then merge with the previous info banner

Author: Not sure about this one. For me this is warning that some of the powertools functionalites cannot be tested locally. in the Info I put different kind of additional information like link to the documentation

Reviewer 2: Found a better place within Tracer -- by the time we add `Tracer` feature it makes this more explicit on what this actually means.

Value               arn:aws:lambda:eu-central-1:123456789012:function:sam-app-HelloWorldFunction-dOcfAtYoEiGo                           
------------------------------------------------------------------------------------------------------------------------------------------
Successfully created/updated stack - sam-app in eu-central-1
```
This command builds a package and deploy it to your AWS Account. You find the API Url against which you can launch requests in the output section. Now, you can trigger your endpoints.
```bash
> curl https://1234567890.execute-api.eu-central-1.amazonaws.com/Prod/hello
{"statusCode":200,"body":{"message":"hello unknown!"}}%
```
!!! Info
    For more details on the SAM deployment mechanism, see [link](https://docs.aws.amazon.com/serverless-application-model/latest/developerguide/sam-cli-command-reference-sam-deploy.html).

## API Gateway router
Let's say we want to have different method that acts like an echo server. It takes user input (username) and output it to the caller. We need to create an API with an URL path `/hello/<name>`, where the `name` string is the input from the user.

> One approach is to create another lambda with required method and set up the API gateway to call it.

Reviewer 2: Not sure I understood what you mean here -- Consider reviewing the wording. Also, use capital `L` for Lambda

Author: Rewrote to this:
```
Let's expand our application with a new method. It takes  an username as a input and return it in the response.

We decided to write another Lambda including required method. Next, we configure our API Gateway to expose this Lambda under new unique path `/hello/{name}`.
```
